In [480]:
import pandas as pd # For data manipulation and analysis using DataFrames
import numpy as np # For numberical computing and array/ matrix operations
import math

In [481]:
dataset1 = pd.read_csv( "https://raw.githubusercontent.com/Zixin-Fan/CSI4142-A2/main/US_Top_50_Universities_2026.csv") # Read the 'US_Top_50_Universities_2026.csv' from a row github url
dataset1.head() # show the first 5 rows of the dataset

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
0,Massachusetts Institute of Technology (MIT),1,1861,Private,MA,100.0,91.6,96.2
1,Columbia University,2,1754,Private,NY,95.9,83.7,92.1
2,Princeton University,3,1746,Private,NJ,99.0,70.0,94.5
3,Stanford University,4,1891,Private,CA,99.5,73.5,97.8
4,"University of California, Berkeley",5,1868,Public,CA,98.9,70.6,91.4


# Validity checker

## Data Type Errors

### Cell 1 – Test name + description
Data Type check:
In this test, we verify the data type of values in the "National_Rank" column. Any non-numeric value is considered a data type error.

### Cell 2 – How you introduced such error 

In [482]:
dataTypeCheckDataset=dataset1.copy() # copy dataset
dataTypeCheckDataset["National_Rank"]=dataTypeCheckDataset["National_Rank"].astype("object") # convert numeric to object in order to insert data type errors

n=math.ceil(len(dataTypeCheckDataset)*0.05) # total number of rows (round up) to modify (5% of the dataset)
errorRows=np.random.choice(dataTypeCheckDataset.index,n,replace=False) # randomly select n rows without duplicates

# .loc selects the "National_Rank" cells for those specific row indices stored in errorRows
dataTypeCheckDataset.loc[errorRows,"National_Rank"]="unknown" # assign "unknown"(object type) to some "National_Rank" cells

### Cell 3 – Checker code

In [483]:
rank_numeric=pd.to_numeric(dataTypeCheckDataset["National_Rank"],errors="coerce") # convert values to numeric; non-numeric entries become NaN
invalid_rank=dataTypeCheckDataset[rank_numeric.isna()] # .isna() detects NaN values, which indicate non-numeric or missing ranks；since the series shares the same index as the dataset, the corresponding rows are returned
invalid_rank # a data frame containing rows with invalid rank (before fix)

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
20,"University of Michigan, Ann Arbor",unknown,1817,Public,MI,93.8,65.8,90.2
40,Georgetown University,unknown,1789,Private,DC,78.4,18.9,90.3
46,Washington and Lee University,unknown,1749,Private,VA,28.5,4.2,91.5


### Cell 4 – Report of findings

In [484]:
invalid_rank # before fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
20,"University of Michigan, Ann Arbor",unknown,1817,Public,MI,93.8,65.8,90.2
40,Georgetown University,unknown,1789,Private,DC,78.4,18.9,90.3
46,Washington and Lee University,unknown,1749,Private,VA,28.5,4.2,91.5


In [485]:
dataTypeCheckDataset.loc[rank_numeric.isna(), "National_Rank"] = np.nan # replace “unknown” in the original column with NaN.
dataTypeCheckDataset[dataTypeCheckDataset["National_Rank"].isna()] # flagged rows after handling

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
20,"University of Michigan, Ann Arbor",NaN,1817,Public,MI,93.8,65.8,90.2
40,Georgetown University,NaN,1789,Private,DC,78.4,18.9,90.3
46,Washington and Lee University,NaN,1749,Private,VA,28.5,4.2,91.5


Since rankings represent ordered positions, filling in values could distort the true ranking structure. Also, because the dataset contains only 50 universities, removing rows would lead to unnecessary data loss. Therefore, invalid ranks were replaced with NaN to preserve data integrity.

## Range errors

In [486]:
min=int(dataset1["Founded_Year"].min())
max=int(dataset1["Founded_Year"].max())
min,max

(1636, 1965)

### Cell 1 – Test name + description
Range check:
In this test, we verify the range of values in the "Founded_Year" column. In the original dataset, founded years range from 1636 to 1965. Based on domain knowledge, a university cannot be founded before 1600 or in the future, so we define the valid range as 1600–2026. Any value outside this range is considered a range error.

### Cell 2 – How you introduced such error 

In [487]:
rangeCheckDataset=dataset1.copy() # copy dataset

n=math.ceil(len(rangeCheckDataset)*0.05) # total number of rows (round up) to modify (5% of the dataset)
errorRows=np.random.choice(rangeCheckDataset.index,n,replace=False) # randomly select n rows without duplicates

half=n//2 # split selected rows into two groups
earlyYear=errorRows[:half] # years that are too early (<1600)
futureYear=errorRows[half:] # years that are in the future (>2026)

# .loc selects the "Founded_Year" cells for those specific row indices stored in earlyYear
# np.random.randint(lower,upper,how many numbers to generate)
rangeCheckDataset.loc[earlyYear,"Founded_Year"]=np.random.randint(1000,1600,len(earlyYear)) # assign early years (1000–1599)
rangeCheckDataset.loc[futureYear,"Founded_Year"]=np.random.randint(2027,3000,len(futureYear)) # assign future years (2027-3000)

### Cell 3 – Checker code

In [488]:
invalid_years=rangeCheckDataset[(rangeCheckDataset["Founded_Year"]<1600)|(rangeCheckDataset["Founded_Year"]>2026)]
invalid_years # a data frame containing rows with invalid years (before fix)

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
36,"University of Maryland, College Park",37,2313,Public,MD,86.4,18.2,82.1
40,Georgetown University,41,2926,Private,DC,78.4,18.9,90.3
42,Carnegie Mellon University,43,1579,Private,PA,99.0,84.3,93.7


### Cell 4 – Report of findings

In [489]:
invalid_years # before fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
36,"University of Maryland, College Park",37,2313,Public,MD,86.4,18.2,82.1
40,Georgetown University,41,2926,Private,DC,78.4,18.9,90.3
42,Carnegie Mellon University,43,1579,Private,PA,99.0,84.3,93.7


In [490]:
rangeCheckDataset.loc[rangeCheckDataset["Founded_Year"]<1600,"Founded_Year"]=1600 # set early years to the lower bound
rangeCheckDataset.loc[rangeCheckDataset["Founded_Year"]>2026,"Founded_Year"]=2026 # set future years to the upper bound
rangeCheckDataset.loc[invalid_years.index] # show the same rows after correction (after fix)

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
36,"University of Maryland, College Park",37,2026,Public,MD,86.4,18.2,82.1
40,Georgetown University,41,2026,Private,DC,78.4,18.9,90.3
42,Carnegie Mellon University,43,1600,Private,PA,99.0,84.3,93.7


The checker identified several invalid values in the "Founded_Year" column, including years earlier than 1600 and years in the future. Values outside the valid range were corrected by replacing them with the nearest boundary (1600 or 2026). This approach preserves the records while ensuring data validity, and imputation was avoided because founding year is a historical fact. After the fix, all values fall within the acceptable range.

## Format errors

### Cell 1 – Test name + description
Format check:
In this test, we verify that values in the "State" column follow the standard two-letter uppercase abbreviation format. Any value that does not match this format is considered a format error.

### Cell 2 – How you introduced such error 

In [491]:
formatCheckDataset=dataset1.copy() # copy dataset

n=math.ceil(len(formatCheckDataset)*0.05) # total number of rows (round up) to modify (5% of the dataset)
errorRows=np.random.choice(formatCheckDataset.index,n,replace=False) # randomly select n rows without duplicates

# .loc selects the "State" cells for those specific row indices stored in errorRows
formatCheckDataset.loc[errorRows,"State"]=" "+formatCheckDataset.loc[errorRows,"State"].str.lower()+"." # e.g. NY -> ny.

### Cell 3 – Checker code

In [492]:
invalid_state=formatCheckDataset[~formatCheckDataset["State"].str.fullmatch(r"[A-Z]{2}")]
invalid_state

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
17,Brown University,18,1764,Private,ri.,88.2,21.3,89.9
38,Bowdoin College,39,1794,Private,me.,30.1,7.2,86.0
45,University of Texas at Austin,46,1883,Public,tx.,90.1,46.2,89.2


### Cell 4 – Report of findings

In [493]:
invalid_state # before fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
17,Brown University,18,1764,Private,ri.,88.2,21.3,89.9
38,Bowdoin College,39,1794,Private,me.,30.1,7.2,86.0
45,University of Texas at Austin,46,1883,Public,tx.,90.1,46.2,89.2


In [494]:
formatCheckDataset["State"]=formatCheckDataset["State"].str.strip().str.upper().str.replace(".","",regex=False)
formatCheckDataset.loc[invalid_state.index] # after fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
17,Brown University,18,1764,Private,RI,88.2,21.3,89.9
38,Bowdoin College,39,1794,Private,ME,30.1,7.2,86.0
45,University of Texas at Austin,46,1883,Public,TX,90.1,46.2,89.2


Format inconsistencies were corrected by standardizing state codes to the two-letter uppercase format. Extra spaces, lowercase letters, and punctuation were removed to ensure consistency.

## Consistency errors

### Cell 1 – Test name + description
Consistency check:
This test checks whether the research impact score is consistent with the national ranking. Universities with top rankings are expected to have high research impact scores. Extremely low values may indicate inconsistent or incorrect data.

### Cell 2 – How you introduced such error 

In [495]:
consistencyCheckDataset=dataset1.copy() # copy dataset

n=math.ceil(len(consistencyCheckDataset)*0.05) # total number of rows (round up) to modify (5% of the dataset)

consistencyCheckDataset.loc[consistencyCheckDataset["National_Rank"]<=n,"Research_Impact_Score"]=0 # the top n schools should have a high research score, but here the research score is 0.

### Cell 3 – Checker code

In [496]:
averageResearchScore=consistencyCheckDataset["Research_Impact_Score"].mean() # calculate the mean of the reserch impact score
consistency_errors = consistencyCheckDataset[(consistencyCheckDataset["National_Rank"]<=n)&(consistencyCheckDataset["Research_Impact_Score"]<averageResearchScore)] # the top n's research scores are lower than the average
consistency_errors

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
0,Massachusetts Institute of Technology (MIT),1,1861,Private,MA,0.0,91.6,96.2
1,Columbia University,2,1754,Private,NY,0.0,83.7,92.1
2,Princeton University,3,1746,Private,NJ,0.0,70.0,94.5


### Cell 4 – Report of findings

In [497]:
consistency_errors # before fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
0,Massachusetts Institute of Technology (MIT),1,1861,Private,MA,0.0,91.6,96.2
1,Columbia University,2,1754,Private,NY,0.0,83.7,92.1
2,Princeton University,3,1746,Private,NJ,0.0,70.0,94.5


In [498]:
consistencyCheckDataset.loc[consistency_errors.index,"Research_Impact_Score"]=dataset1.loc[consistency_errors.index,"Research_Impact_Score"]

consistencyCheckDataset.loc[consistency_errors.index] # after fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
0,Massachusetts Institute of Technology (MIT),1,1861,Private,MA,100.0,91.6,96.2
1,Columbia University,2,1754,Private,NY,95.9,83.7,92.1
2,Princeton University,3,1746,Private,NJ,99.0,70.0,94.5


We used the dataset’s average research impact score as a baseline threshold to flag unusually low values among top-ranked universities. The detected inconsistencies were corrected by restoring the original research impact scores. The rows above show the corrected values.

## Uniqueness error

### Cell 1 – Test name + description
Uniqueness check:
This test verifies that each university name appears only once. Duplicate names indicate a uniqueness error.

### Cell 2 – How you introduced such error 

In [499]:
uniquenessCheckDataset=dataset1.copy() # copy dataset

n=math.ceil(len(uniquenessCheckDataset)*0.05) # total number of rows (round up) to modify (5% of the dataset)
errorRows=np.random.choice(uniquenessCheckDataset.index,n,replace=False) # randomly select n rows without duplicates

duplicate_name=uniquenessCheckDataset.loc[errorRows[0], "University_Name"] # the first university name in errorRows
# .loc selects the "University_Name" cells for those specific row indices stored in errorRows
uniquenessCheckDataset.loc[errorRows,"University_Name"]=duplicate_name

### Cell 3 – Checker code

In [500]:
duplicate_universities=uniquenessCheckDataset[uniquenessCheckDataset.duplicated(subset="University_Name", keep=False)] # checks duplicates based only on the university name and keep=False marks ALL occurrences
duplicate_universities

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
14,Brown University,15,1919,Public,CA,95.6,69.2,89.4
17,Brown University,18,1764,Private,RI,88.2,21.3,89.9
23,Brown University,24,1891,Private,CA,96.8,90.7,95.1


### Cell 4 – Report of findings

In [501]:
duplicate_universities # before fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
14,Brown University,15,1919,Public,CA,95.6,69.2,89.4
17,Brown University,18,1764,Private,RI,88.2,21.3,89.9
23,Brown University,24,1891,Private,CA,96.8,90.7,95.1


In [502]:
uniquenessCheckDataset.loc[duplicate_universities.index, "University_Name"] = dataset1.loc[duplicate_universities.index, "University_Name"] # restore original university names
uniquenessCheckDataset.loc[duplicate_universities.index] # after fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
14,"University of California, Los Angeles (UCLA)",15,1919,Public,CA,95.6,69.2,89.4
17,Brown University,18,1764,Private,RI,88.2,21.3,89.9
23,California Institute of Technology (Caltech),24,1891,Private,CA,96.8,90.7,95.1


Duplicate university names were detected. Since the dataset originally contained unique entries and includes only 50 institutions, removing rows would cause unnecessary data loss. Therefore, the original names were restored to preserve data integrity and maintain one unique record per university.

## Presence errors

### Cell 1 – Test name + description
Presence check:
This test verifies whether values exist in the "Employment_Rate" column. Missing values are considered presence errors because employment rate is a key performance indicator.

### Cell 2 – How you introduced such error 

In [503]:
presenceCheckDataset=dataset1.copy() # copy dataset

n=math.ceil(len(presenceCheckDataset)*0.05) # total number of rows (round up) to modify (5% of the dataset)
errorRows=np.random.choice(presenceCheckDataset.index,n,replace=False) # randomly select n rows without duplicates

# .loc selects the "Employment_Rate" cells for those specific row indices stored in errorRows
presenceCheckDataset.loc[errorRows,"Employment_Rate"]=np.nan

### Cell 3 – Checker code

In [504]:
invalid_employment_rate=presenceCheckDataset[presenceCheckDataset["Employment_Rate"].isna()] # select rows where employment rate is NaN
invalid_employment_rate

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
30,"University of California, Irvine",31,1965,Public,CA,89.1,21.4,NaN
36,"University of Maryland, College Park",37,1856,Public,MD,86.4,18.2,NaN
37,"University of Illinois, Urbana-Champaign",38,1867,Public,IL,85.3,62.3,NaN


### Cell 4 – Report of findings

In [505]:
invalid_employment_rate # before fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
30,"University of California, Irvine",31,1965,Public,CA,89.1,21.4,NaN
36,"University of Maryland, College Park",37,1856,Public,MD,86.4,18.2,NaN
37,"University of Illinois, Urbana-Champaign",38,1867,Public,IL,85.3,62.3,NaN


In [506]:
presenceCheckDataset.loc[invalid_employment_rate.index,"Employment_Rate"]=dataset1.loc[invalid_employment_rate.index,"Employment_Rate"]
presenceCheckDataset.loc[invalid_employment_rate.index] # after fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
30,"University of California, Irvine",31,1965,Public,CA,89.1,21.4,83.5
36,"University of Maryland, College Park",37,1856,Public,MD,86.4,18.2,82.1
37,"University of Illinois, Urbana-Champaign",38,1867,Public,IL,85.3,62.3,87.5


Missing values were detected in the Employment_Rate column. Since this field is essential for performance analysis, the original values were restored to preserve data integrity.

## Length Errors

### Cell 1 – Test name + description
Length check:
This test verifies that Intl_Student_Ratio should have 1 decimal place.

### Cell 2 – How you introduced such error 

In [507]:
lengthCheckDataset=dataset1.copy() # copy dataset

n=math.ceil(len(lengthCheckDataset)*0.05) # total number of rows (round up) to modify (5% of the dataset)
errorRows=np.random.choice(lengthCheckDataset.index,n,replace=False) # randomly select n rows without duplicates

# .loc selects the "Intl_Student_Ratio" cells for those specific row indices stored in errorRows
lengthCheckDataset.loc[errorRows,"Intl_Student_Ratio"]=lengthCheckDataset.loc[errorRows,"Intl_Student_Ratio"]+0.011111111

### Cell 3 – Checker code

In [508]:
invalid_length=lengthCheckDataset[lengthCheckDataset["Intl_Student_Ratio"].astype(str).str.split(".").str[1].fillna("").str.len()>1] # transfer to string -> split by . -> get the second part -> fill missing values (if no decimal part) -> length of decimals
invalid_length

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
4,"University of California, Berkeley",5,1868,Public,CA,98.9,70.611111,91.4
39,Brigham Young University,40,1875,Private,UT,45.2,5.111111,84.4
42,Carnegie Mellon University,43,1900,Private,PA,99.0,84.311111,93.7


### Cell 4 – Report of findings

In [509]:
invalid_length # before fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
4,"University of California, Berkeley",5,1868,Public,CA,98.9,70.611111,91.4
39,Brigham Young University,40,1875,Private,UT,45.2,5.111111,84.4
42,Carnegie Mellon University,43,1900,Private,PA,99.0,84.311111,93.7


In [510]:
lengthCheckDataset.loc[invalid_length.index,"Intl_Student_Ratio"]=lengthCheckDataset.loc[invalid_length.index,"Intl_Student_Ratio"].round(1)
lengthCheckDataset.loc[invalid_length.index] # after fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
4,"University of California, Berkeley",5,1868,Public,CA,98.9,70.6,91.4
39,Brigham Young University,40,1875,Private,UT,45.2,5.1,84.4
42,Carnegie Mellon University,43,1900,Private,PA,99.0,84.3,93.7


The checker detected values in "Intl_Student_Ratio" with more than one digit after the decimal point, which violates the dataset’s consistent 1-decimal format. To solve the issue, all flagged values were rounded to 1 decimal place to standardize precision.

## Look-up errors

### Cell 1 – Test name + description
Look-up check:
This test verifies that "Institution_Type" belongs to a predefined set of valid categories: public and private.

### Cell 2 – How you introduced such error 

In [511]:
lookupCheckDataset=dataset1.copy() # copy dataset

n=math.ceil(len(lookupCheckDataset)*0.05) # total number of rows (round up) to modify (5% of the dataset)
errorRows=np.random.choice(lookupCheckDataset.index,n,replace=False) # randomly select n rows without duplicates

# .loc selects the "Institution_Type" cells for those specific row indices stored in errorRows
lookupCheckDataset.loc[errorRows,"Institution_Type"]="NA"

### Cell 3 – Checker code

In [512]:
valid_types=["Public", "Private"] # a list of valid values
invalid_lookup=lookupCheckDataset[~lookupCheckDataset["Institution_Type"].isin(valid_types)] # test whether the values are in the valid set
invalid_lookup

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
20,"University of Michigan, Ann Arbor",21,1817,NA,MI,93.8,65.8,90.2
38,Bowdoin College,39,1794,NA,ME,30.1,7.2,86.0
45,University of Texas at Austin,46,1883,NA,TX,90.1,46.2,89.2


### Cell 4 – Report of findings

In [513]:
invalid_lookup # before fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
20,"University of Michigan, Ann Arbor",21,1817,NA,MI,93.8,65.8,90.2
38,Bowdoin College,39,1794,NA,ME,30.1,7.2,86.0
45,University of Texas at Austin,46,1883,NA,TX,90.1,46.2,89.2


In [514]:
lookupCheckDataset.loc[invalid_lookup.index,"Institution_Type"]=dataset1.loc[invalid_lookup.index,"Institution_Type"]
lookupCheckDataset.loc[invalid_lookup.index] # after fix

,University_Name,National_Rank,Founded_Year,Institution_Type,State,Research_Impact_Score,Intl_Student_Ratio,Employment_Rate
20,"University of Michigan, Ann Arbor",21,1817,Public,MI,93.8,65.8,90.2
38,Bowdoin College,39,1794,Private,ME,30.1,7.2,86.0
45,University of Texas at Austin,46,1883,Public,TX,90.1,46.2,89.2


Invalid values (NA) were detected in the "Institution_Type" column, which is not part of the predefined categories: public and private. The incorrect entries were restored using the original dataset to ensure consistent categorical values.

# AI Assistance Disclosure
- understand how pandas.to_numeric(errors="coerce") detects non-numeric values
- learn how to randomly select rows using NumPy
- understand how '.loc' assigns values to selected rows
- understand how the pandas '.isin()' function can validate values against a predefined set
- suggest simple strategies for introducing and correcting errors
- format expressions for format error checks
- detect duplicate values for uniqueness checks